sympy/solvers/ode.py

```diff
@@ -346,6 +346,160 @@
+            b: dep.coeff(f(x)),
+            c: ind}
+
+        if not r[a].has(f) and not r[b].has(f) and (
+                r[a]*df + r[b]*f(x) + r[c]).expand() - reduced_eq == 0:
+            r['a'] = a
+            r['b'] = b
+            r['c'] = c
+
+        return r
+
+    def match(self):
+        # return True if this method can solve the equation
+        if self.order == 1:
+            r = self._get_coeffs_linear_ode()
+            if r is not None and r['a'] != 0:
+                return True
+        return False
+
+    def general_solution(self):
+        r"""
+        Solves 1st order linear differential equations.
+
+        These are differential equations of the form
+
+        .. math:: dy/dx + P(x) y = Q(x)\text{.}
+
+        These kinds of differential equations can be solved in a general way.  The
+        integrating factor `e^{\int P(x) \,dx}` will turn the equation into a
+        separable equation.  The general solution is::
+
+            >>> from sympy import Function, _dsolve, Eq, pprint, diff, sin
+            >>> from sympy.abc import x
+            >>> f, P, Q = map(Function, ['f', 'P', 'Q'])
+            >>> genform = Eq(f(x).diff(x) + P(x)*f(x), Q(x))
+            >>> pprint(genform)
+                        d
+            P(x)*f(x) + --(f(x)) = Q(x)
+                        dx
+            >>> pprint(_dsolve(genform, f(x), hint='1st_linear_Integral'))
+                /       /                   \
+                |      |                    |
+                |      |         /          |     /
+                |      |        |           |    |
+                |      |        | P(x) dx   |  - | P(x) dx
+                |      |        |           |    |
+                |      |       /            |   /
+            f(x) = |C1 +  | Q(x)*e           dx|*e
+                |      |                    |
+                \     /                     /
+
+
+        Examples
+        ========
+
+        >>> f = Function('f')
+        >>> pprint(_dsolve(Eq(x*diff(f(x), x) - f(x), x**2*sin(x)),
+        ... f(x), '1st_linear'))
+        f(x) = x*(C1 - cos(x))
+
+        References
+        ==========
+
+        - https://en.wikipedia.org/wiki/Linear_differential_equation#First_order_equation
+        - M. Tenenbaum & H. Pollard, "Ordinary Differential Equations",
+        Dover 1963, pp. 92
+
+        """
+        x = self.func.args[0]
+        f = self.func.func
+        r = self._get_coeffs_linear_ode()  # a*diff(f(x),x) + b*f(x) + c
+        f = r.get('u', f(x))  # take almost-linear u if present, else f(x)
+        hints = {'default': '1st_linear',
+                '1st_linear': r,
+                'order': 1,
+                'hint': '1st_linear',
+                'func': f,
+                }
+        return _helper_simplify(self.eq, hints['hint'], hints, self.simplify, self.ics)
+
+all_solvers = [FirstLinear]
+
+def _dsolve(eq, func=None, hint="default", simplify=True,
+    ics=None, **kwargs):
```

Comment:
We don't want a new function `_dsolve`. We need the original `dsolve` to use the `FirstLinear` class for matching and solving the ODE.

Reply:
I thought we'd put things together in this function and then replace dsolve with it.

Reply:
The problem with that is that we can't test the function as the changes are made. We need to do this in small steps ensuring that the tests pass at each point (see #18403)

Reply:
>  (see #18403)

Ok then, I'll make the changes in the above format. Should I close this PR for now?

Reply:
Yes, I think this one can be closed.